crates/sui-core/src/authority_active/checkpoint_driver/mod.rs

```diff
@@ -0,0 +1,551 @@
+            // Our work here is done, we have progressed past the checkpoint for which we were given a proposal.
+            // Our DB has been updated (presumably by consensus) with the sought information (a checkpoint
+            // for this sequence number)
+            break;
+        }
+
+        // We have ran out of authorities?
+        if available_authorities.is_empty() {
+            // We have created as many fragments as possible, so exit.
+            break;
+        }
+
+        let random_authority = _active_authority.state.committee.sample();
+        if available_authorities.remove(random_authority) {
+            // Get a client
+            let client = _active_authority.net.authority_clients[random_authority].clone();
+
+            if let Ok(response) = client
+                .handle_checkpoint(CheckpointRequest::latest(true))
+                .await
+            {
+                if let AuthorityCheckpointInfo::Proposal { current, previous } = &response.info {
+                    // Check if there is a latest checkpoint
+                    if let AuthenticatedCheckpoint::Certified(prev) = previous {
+                        if prev.checkpoint.sequence_number > next_checkpoint_sequence_number {
+                            // We are now way behind, return
+                            return;
+                        }
+                    }
+
+                    // For some reason the proposal is empty?
+                    if current.is_none() || response.detail.is_none() {
+                        continue;
+                    }
+
+                    // TODO: check the proposal is also for the same checkpoint sequence number?
+                    if current.as_ref().unwrap().0.checkpoint.sequence_number()
+                        != _my_proposal.sequence_number()
+                    {
+                        return;
+                    }
+
+                    let other_proposal = CheckpointProposal::new(
+                        current.as_ref().unwrap().clone(),
+                        response.detail.unwrap(),
+                    );
+
+                    let fragment = _my_proposal.fragment_with(&other_proposal);
+                    let _ = checkpoint_db.lock().handle_receive_fragment(&fragment);
+
+                    // TODO: here we should really wait until the fragment is sequenced, otherwise
+                    //       we would be going ahead and sequencing more fragments that may not be
+                    //       needed. For the moment we just linearly back-off.
+                    fragments_num += 1;
+                    if fragments_num > 2 {
+                        tokio::time::sleep(Duration::from_secs(3 * fragments_num)).await;
+                    }
+                }
+            } else {
+                continue;
+            }
+        } else {
+            tokio::time::sleep(Duration::from_millis(100)).await;
```

Comment:
Can this happen? You already checked there are still available authorities at line 493

Reply:
Yeah, I deleted this one